File: backend/src/main/java/com/springboot/controller/LifeguardLocationController.java
package com.springboot.controller;

import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.springboot.annotation.AuthCheck;
import com.springboot.common.BaseResponse;
import com.springboot.common.DeleteRequest;
import com.springboot.common.ErrorCode;
import com.springboot.common.ResultUtils;
import com.springboot.constant.RoleConstant;
import com.springboot.exception.BusinessException;
import com.springboot.exception.ThrowUtils;
import com.springboot.model.dto.lifeguardlocationlog.LifeguardLocationLogAddRequest;
import com.springboot.model.dto.lifeguardlocationlog.LifeguardLocationLogEditRequest;
import com.springboot.model.dto.lifeguardlocationlog.LifeguardLocationLogQueryRequest;
import com.springboot.model.dto.lifeguardlocationlog.LifeguardLocationLogUpdateRequest;
import com.springboot.model.entity.LifeguardLocationLog;
import com.springboot.model.vo.LifeguardLocationLogVO;
import com.springboot.service.LifeguardLocationLogService;
import jakarta.annotation.Resource;
import java.util.List;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/lifeguards/location-logs")
public class LifeguardLocationController {

    @Resource
    private LifeguardLocationLogService lifeguardLocationLogService;

    @PostMapping("/add")
    @AuthCheck(mustRole = RoleConstant.VENUE_ADMIN)
    public BaseResponse<Long> addLifeguardLocationLog(@RequestBody LifeguardLocationLogAddRequest lifeguardLocationLogAddRequest) {
        if (lifeguardLocationLogAddRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        LifeguardLocationLog lifeguardLocationLog = toLifeguardLocationLog(lifeguardLocationLogAddRequest);
        lifeguardLocationLogService.validLifeguardLocationLog(lifeguardLocationLog, true);
        boolean result = lifeguardLocationLogService.save(lifeguardLocationLog);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(lifeguardLocationLog.getId());
    }

    @PostMapping("/delete")
    @AuthCheck(mustRole = RoleConstant.VENUE_ADMIN)
    public BaseResponse<Boolean> deleteLifeguardLocationLog(@RequestBody DeleteRequest deleteRequest) {
        if (deleteRequest == null || deleteRequest.getId() == null || deleteRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        boolean result = lifeguardLocationLogService.removeById(deleteRequest.getId());
        return ResultUtils.success(result);
    }

    @PostMapping("/update")
    @AuthCheck(mustRole = RoleConstant.VENUE_ADMIN)
    public BaseResponse<Boolean> updateLifeguardLocationLog(
            @RequestBody LifeguardLocationLogUpdateRequest lifeguardLocationLogUpdateRequest) {
        if (lifeguardLocationLogUpdateRequest == null || lifeguardLocationLogUpdateRequest.getId() == null
                || lifeguardLocationLogUpdateRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        LifeguardLocationLog lifeguardLocationLog = toLifeguardLocationLog(lifeguardLocationLogUpdateRequest);
        lifeguardLocationLogService.validLifeguardLocationLog(lifeguardLocationLog, false);
        boolean result = lifeguardLocationLogService.updateById(lifeguardLocationLog);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(true);
    }

    @PostMapping("/edit")
    @AuthCheck(mustRole = RoleConstant.VENUE_ADMIN)
    public BaseResponse<Boolean> editLifeguardLocationLog(@RequestBody LifeguardLocationLogEditRequest lifeguardLocationLogEditRequest) {
        if (lifeguardLocationLogEditRequest == null || lifeguardLocationLogEditRequest.getId() == null
                || lifeguardLocationLogEditRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        LifeguardLocationLog lifeguardLocationLog = new LifeguardLocationLog();
        lifeguardLocationLog.setId(lifeguardLocationLogEditRequest.getId());
        lifeguardLocationLog.setLongitude(lifeguardLocationLogEditRequest.getLongitude());
        lifeguardLocationLog.setLatitude(lifeguardLocationLogEditRequest.getLatitude());
        lifeguardLocationLog.setIn_fence(lifeguardLocationLogEditRequest.getInFence());
        lifeguardLocationLog.setReport_source(lifeguardLocationLogEditRequest.getReportSource());
        lifeguardLocationLog.setReported_at(lifeguardLocationLogEditRequest.getReportedAt());
        lifeguardLocationLogService.validLifeguardLocationLog(lifeguardLocationLog, false);
        boolean result = lifeguardLocationLogService.updateById(lifeguardLocationLog);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return ResultUtils.success(true);
    }

    @GetMapping("/get")
    public BaseResponse<LifeguardLocationLog> getLifeguardLocationLogById(long id) {
        if (id <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        LifeguardLocationLog lifeguardLocationLog = lifeguardLocationLogService.getById(id);
        ThrowUtils.throwIf(lifeguardLocationLog == null, ErrorCode.NOT_FOUND_ERROR);
        return ResultUtils.success(lifeguardLocationLog);
    }

    @GetMapping("/get/vo")
    public BaseResponse<LifeguardLocationLogVO> getLifeguardLocationLogVOById(long id) {
        BaseResponse<LifeguardLocationLog> response = getLifeguardLocationLogById(id);
        return ResultUtils.success(lifeguardLocationLogService.getLifeguardLocationLogVO(response.getData()));
    }

    @PostMapping("/list/page")
    public BaseResponse<Page<LifeguardLocationLog>> listLifeguardLocationLogByPage(
            @RequestBody LifeguardLocationLogQueryRequest lifeguardLocationLogQueryRequest) {
        if (lifeguardLocationLogQueryRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        long current = lifeguardLocationLogQueryRequest.getCurrent();
        long size = lifeguardLocationLogQueryRequest.getPageSize();
        ThrowUtils.throwIf(size > 100, ErrorCode.PARAMS_ERROR, "分页大小不能超过100");
        Page<LifeguardLocationLog> lifeguardLocationLogPage = lifeguardLocationLogService.page(new Page<>(current, size),
                lifeguardLocationLogService.getQueryWrapper(lifeguardLocationLogQueryRequest));
        return ResultUtils.success(lifeguardLocationLogPage);
    }

    @PostMapping("/list/page/vo")
    public BaseResponse<Page<LifeguardLocationLogVO>> listLifeguardLocationLogVOByPage(
            @RequestBody LifeguardLocationLogQueryRequest lifeguardLocationLogQueryRequest) {
        BaseResponse<Page<LifeguardLocationLog>> response = listLifeguardLocationLogByPage(lifeguardLocationLogQueryRequest);
        Page<LifeguardLocationLog> lifeguardLocationLogPage = response.getData();
        Page<LifeguardLocationLogVO> lifeguardLocationLogVOPage = new Page<>(lifeguardLocationLogPage.getCurrent(),
                lifeguardLocationLogPage.getSize(), lifeguardLocationLogPage.getTotal());
        List<LifeguardLocationLogVO> lifeguardLocationLogVOList = lifeguardLocationLogService.getLifeguardLocationLogVO(
                lifeguardLocationLogPage.getRecords());
        lifeguardLocationLogVOPage.setRecords(lifeguardLocationLogVOList);
        return ResultUtils.success(lifeguardLocationLogVOPage);
    }

    private LifeguardLocationLog toLifeguardLocationLog(LifeguardLocationLogAddRequest request) {
        LifeguardLocationLog lifeguardLocationLog = new LifeguardLocationLog();
        lifeguardLocationLog.setLifeguard_id(request.getLifeguardId());
        lifeguardLocationLog.setVenue_id(request.getVenueId());
        lifeguardLocationLog.setLongitude(request.getLongitude());
        lifeguardLocationLog.setLatitude(request.getLatitude());
        lifeguardLocationLog.setIn_fence(request.getInFence());
        lifeguardLocationLog.setReport_source(request.getReportSource());
        lifeguardLocationLog.setReported_at(request.getReportedAt());
        return lifeguardLocationLog;
    }

    private LifeguardLocationLog toLifeguardLocationLog(LifeguardLocationLogUpdateRequest request) {
        LifeguardLocationLog lifeguardLocationLog = new LifeguardLocationLog();
        lifeguardLocationLog.setId(request.getId());
        lifeguardLocationLog.setLifeguard_id(request.getLifeguardId());
        lifeguardLocationLog.setVenue_id(request.getVenueId());
        lifeguardLocationLog.setLongitude(request.getLongitude());
        lifeguardLocationLog.setLatitude(request.getLatitude());
        lifeguardLocationLog.setIn_fence(request.getInFence());
        lifeguardLocationLog.setReport_source(request.getReportSource());
        lifeguardLocationLog.setReported_at(request.getReportedAt());
        return lifeguardLocationLog;
    }
}

File: frontend/src/api/lifeguardLocationController.ts
// @ts-ignore
/* eslint-disable */
import request from "@/request";

/** 此处后端没有提供注释 POST /lifeguards/location-logs/add */
export async function addLifeguardLocationLog(
  body: API.LifeguardLocationLogAddRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseLong>("/lifeguards/location-logs/add", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** 此处后端没有提供注释 POST /lifeguards/location-logs/delete */
export async function deleteLifeguardLocationLog(
  body: API.DeleteRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>("/lifeguards/location-logs/delete", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** 此处后端没有提供注释 POST /lifeguards/location-logs/edit */
export async function editLifeguardLocationLog(
  body: API.LifeguardLocationLogEditRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>("/lifeguards/location-logs/edit", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** 此处后端没有提供注释 GET /lifeguards/location-logs/get */
export async function getLifeguardLocationLogById(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getLifeguardLocationLogByIdParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseLifeguardLocationLog>(
    "/lifeguards/location-logs/get",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 GET /lifeguards/location-logs/get/vo */
export async function getLifeguardLocationLogVoById(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getLifeguardLocationLogVOByIdParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseLifeguardLocationLogVO>(
    "/lifeguards/location-logs/get/vo",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /lifeguards/location-logs/list/page */
export async function listLifeguardLocationLogByPage(
  body: API.LifeguardLocationLogQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageLifeguardLocationLog>(
    "/lifeguards/location-logs/list/page",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /lifeguards/location-logs/list/page/vo */
export async function listLifeguardLocationLogVoByPage(
  body: API.LifeguardLocationLogQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageLifeguardLocationLogVO>(
    "/lifeguards/location-logs/list/page/vo",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /lifeguards/location-logs/update */
export async function updateLifeguardLocationLog(
  body: API.LifeguardLocationLogUpdateRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>("/lifeguards/location-logs/update", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}
